go/mysql/binlog_event_compression.go

```diff
@@ -192,78 +211,118 @@ func (tp *TransactionPayload) read(data []byte) error {
-		err               error
-	)
-
-	// Switch to slower but less memory intensive stream mode for larger payloads.
-	if tp.UncompressedSize > zstdInMemoryDecompressorMaxSize {
-		in := bytes.NewReader(tp.Payload)
-		streamDecoder, err := zstd.NewReader(in)
-		if err != nil {
-			return nil, err
-		}
-		defer streamDecoder.Close()
-		out := io.Writer(&bytes.Buffer{})
-		_, err = io.Copy(out, streamDecoder)
-		if err != nil {
-			return nil, err
-		}
-		decompressedBytes = out.(*bytes.Buffer).Bytes()
-	} else { // Process smaller payloads using in-memory buffers.
-		decompressedBytes, err = zstdDecoder.DecodeAll(tp.Payload, nil)
+
+	// Switch to slower but less memory intensive stream mode for
+	// larger payloads.
+	if tp.uncompressedSize > zstdInMemoryDecompressorMaxSize {
+		in := bytes.NewReader(tp.payload)
+		streamDecoder, err := zstd.NewReader(in, zstd.WithDecoderMaxMemory(zstdInMemoryDecompressorMaxSize))
 		if err != nil {
-			return nil, err
+			return err
 		}
+		compressedTrxPayloadsUsingStream.Add(1)
+		tp.reader = streamDecoder.IOReadCloser()
+		return nil
+	}
+
+	// Process smaller payloads using only in-memory buffers.
+	decompressedBytes := make([]byte, 0, tp.uncompressedSize) // Perform a single pre-allocation
+	decompressedBytes, err := zstdDecoder.DecodeAll(tp.payload, decompressedBytes[:0])
+	if err != nil {
+		return err
 	}
+	if uint64(len(decompressedBytes)) != tp.uncompressedSize {
+		return vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT,
+			"uncompressed transaction payload size %d does not match expected size %d", len(decompressedBytes), tp.uncompressedSize)
+	}
+	compressedTrxPayloadsInMem.Add(1)
+	tp.reader = bytes.NewReader(decompressedBytes)
+	return nil
+}
 
-	if uint64(len(decompressedBytes)) != tp.UncompressedSize {
-		return []byte{}, vterrors.New(vtrpcpb.Code_INVALID_ARGUMENT,
-			fmt.Sprintf("decompressed size %d does not match expected size %d", len(decompressedBytes), tp.UncompressedSize))
+// Close should be called in a defer where the TransactionPayload is
+// used to ensure that the underlying reader and related resources
+// used are cleaned up.
+func (tp *TransactionPayload) Close() {
+	switch reader := tp.reader.(type) {
+	case *bytes.Reader:
+		reader = nil
+	case io.ReadCloser:
+		reader.Close()
 	}
+	tp.iterator = nil
+}
 
-	return decompressedBytes, nil
+// GetNextEvent returns the next binlog event that was contained within
+// the compressed transaction payload. It will return io.EOF when there
+// are no more events left in the payload.
+func (tp *TransactionPayload) GetNextEvent() (BinlogEvent, error) {
+	if tp == nil || tp.iterator == nil {
+		return nil, vterrors.New(vtrpcpb.Code_INVALID_ARGUMENT, "TransactionPayload has been closed")
+	}
+	return tp.iterator()
 }
+
+// Events returns an iterator over the internal binlog events that
+// were contained within the compressed transaction payload/event.
+// It returns a single-use iterator.
+// TODO: come back to this when main is on go 1.23. See:
```

Comment:
Although we can git blame or investigate on GitHub, please add a link to this PR; preferably also change to `TODO(mattlord)`.